File: 04 April/04-16/string_math.py
def do_math(s):
    array = [int(match[0]) if is_integer(match[0]) else match[0] for match in re.finditer(r"\d+|[^\d]+", s)]
    result = array[0] if is_integer(array[0]) else array[1]

    for index, element in enumerate(array):
        if not is_integer(element):
            if (index != 0 and index != len(array) - 1):
                result += (-1 if len(element) % 2 else 1) * array[index + 1]

    return result

def is_integer(element):
    try:
        int(element)
    except (ValueError, TypeError):
        return False
    else:
        return True

# --- TEST SUITE ---

tests_text = r'''
1. do_math("3ab10c8") should return 5.
2. do_math("6MINUS4") should return 2.
3. do_math("9plus3") should return 12.
4. do_math("5fkwo#10i#%.<>15P=@20!#B/25") should return 15.
5. do_math("a.67,1$lk6ldf34@#LD@]2d32d2'2l3,@l3L#@2gh35s09if=df#$t9sm49t0df3$^%[vc;:0:4mt") should return 67.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 04 April/04-16/stringMath.js
function doMath(str) {
  const array = [...str.matchAll(/\d+|[^\d]+/g)]
    .map(match => isInteger(match[0]) ? parseInt(match[0]) : match[0]);
  let result = isInteger(array[0]) ? array[0] : array[1];

  array.forEach((match, index, source) => {
    if (!isInteger(match)) {
      if (index !== 0 && index !== source.length - 1) {
        result += (match.length % 2 ? -1 : 1) * source[index + 1];
      }
    }
  })

  return result;
}

const isInteger = element => !Number.isNaN(parseInt(element));

// --- TEST SUITE ---

const testsText = String.raw`
1. doMath("3ab10c8") should return 5.
2. doMath("6MINUS4") should return 2.
3. doMath("9plus3") should return 12.
4. doMath("5fkwo#10i#%.<>15P=@20!#B/25") should return 15.
5. doMath("a.67,1$lk6ldf34@#LD@]2d32d2'2l3,@l3L#@2gh35s09if=df#$t9sm49t0df3$^%[vc;:0:4mt") should return 67.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length) return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i])) {
          return false;
        }
      } else {
        return false;
      }
    } else if (a[i] !== b[i]) {
      return false;
    }
  }

  return true;
}

runTests(testData);